### PR TITLE
feat: Add gradle plugin

### PR DESCRIPTION
**Pull request by caalador**
Add gradle plugin into the plugins-module.

### CODE REVIEW

File: flow-plugins/flow-gradle-plugin/README.md [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+
+| Vaadin Gradle Plugin version | Supports |
+|------------------------------|----------|
+| -                            | Vaadin 13 and lower are unsupported |
+| 0.6.0 and lower              | Vaadin 14.1.x LTS and lower |
+| 0.7.0                        | Vaadin 14.2.x LTS |
+| 0.14.3.7                     | All other Vaadin 14 LTS versions (recommended) |
+| -                            | Vaadin 15 and 16 are unsupported |
+| 0.17.0.1                     | Vaadin 17 and higher |
+| 0.20.0.0.alpha3              | Vaadin 20 and higher, experimental support for Vaadin 19 |
```
Comment (pleku) [inline flow-plugins/flow-gradle-plugin/README.md:45]:
```suggestion
| 0.20.0.0.alpha3              | experimental support for Vaadin 19 and 20 |
| 20.0.0                            | Vaadin 20 and higher |
```

File: flow-plugins/flow-gradle-plugin/README.md [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+   then packages everything into the war/jar archive. Automatically calls the
+   `vaadinPrepareFrontend` and `vaadinBuildFrontend` tasks.
+
+*Note* (after you built the project in production mode): In order to prepare the project
+setup back to development mode, you must run `./gradlew vaadinPrepareFrontend`
+with the `productionMode` effectively set to false (e.g. by ommitting the `-Pvaadin.productionMode` flag).
+
+All tasks provided by the plugin:
+
+* `vaadinClean` will clean the project completely and removes JavaScript packaging-related
```
Comment (pleku) [inline flow-plugins/flow-gradle-plugin/README.md:63]: Just commenting that we don't have this for the maven plugin but should make it unified.

File: flow-plugins/flow-gradle-plugin/README.md [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+  need to be invoked explicitly. Note: this task is only triggered automatically if `productionMode` is set to true.
+
+## Configuration
+
+To configure the plugin, you can use the following snippet in your `build.gradle` file:
+
+`build.gradle` in Groovy:
+```groovy
+vaadin {
+  pnpmEnable = true
```
Comment (pleku) [inline flow-plugins/flow-gradle-plugin/README.md:84]: I don't get this - how does `pnpmEnable` relate to enabling vaadin gradle plugin? Or is it surplus ? Should it be production mode instead ?

File: flow-plugins/flow-gradle-plugin/README.md [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+}
+```
+
+All configuration options follow. Note that you **RARELY** need to change anything of the below.
```
Comment (pleku) [inline flow-plugins/flow-gradle-plugin/README.md:88]: I disagree, production mode is something that you should be aware of.

File: flow-plugins/flow-gradle-plugin/README.md [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+then produce the `META-INF/VAADIN/config/flow-build-info.json` file with the `"productionMode": true` setting,
+which will then cause Vaadin to start in production mode at runtime. To revert this setting
+back to false, simply recompile your project without the `-Pvaadin.productionMode` switch.
+
+# Developing The Plugin
+
+Please read the Gradle Tutorial on [Developing Custom Gradle Plugins](https://docs.gradle.org/current/userguide/custom_plugins.html)
+to understand how Gradle plugins are developed.
+
+The main entry to the plugin is the `VaadinPlugin` class. When applied to the project, it will register
```
Comment (pleku) [inline flow-plugins/flow-gradle-plugin/README.md:217]: I wonder if this is still valid, will see soon

File: flow-plugins/flow-gradle-plugin/README.md [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+all necessary tasks and extensions into the project.
+
+Use Intellij (Community is enough) to open the project.
+
+## Branches
```
Comment (pleku) [inline flow-plugins/flow-gradle-plugin/README.md:222]: This is not up to date, should just remove it

File: flow-plugins/flow-gradle-plugin/pom.xml [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+    <parent>
+        <groupId>com.vaadin</groupId>
+        <artifactId>flow-plugins</artifactId>
+        <version>7.0-SNAPSHOT</version>
+    </parent>
+
+    <artifactId>flow-gradle-plugin</artifactId>
+    <name>Flow Gradle Plugin</name>
+    <packaging>jar</packaging>
+
```
Comment (pleku) [inline flow-plugins/flow-gradle-plugin/pom.xml:12]: Description missing - please add with information on how this relates to `vaadin-gradle-plugin`

File: flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/VaadinBuildFrontendTask.kt [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+ * annotations defined in the classpath,
+ * * Copy resource files used by flow from `.jar` files to the `node_modules`
+ * folder
+ * * Install dependencies by running `npm install`
+ * * Update the [FrontendUtils.IMPORTS_NAME] file imports with the
+ * [com.vaadin.flow.component.dependency.JsModule] [com.vaadin.flow.theme.Theme] and [com.vaadin.flow.component.dependency.JavaScript] annotations defined in
+ * the classpath,
+ * * Update [FrontendUtils.WEBPACK_CONFIG] file.
+ *
+ * @since 2.0
```
Comment (pleku) [inline flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/VaadinBuildFrontendTask.kt:39]:
```suggestion
```

File: flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/VaadinFlowPluginExtension.kt [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+import com.vaadin.flow.server.InitParameters
+import com.vaadin.flow.server.frontend.FrontendTools
+import com.vaadin.flow.server.frontend.FrontendUtils
+import com.vaadin.flow.server.frontend.installer.NodeInstaller
+import org.gradle.api.Project
+import org.gradle.api.plugins.JavaPluginConvention
+import org.gradle.api.tasks.SourceSetContainer
+import java.io.File
+
+public open class VaadinFlowPluginExtension(project: Project) {
```
Comment (pleku) [inline flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/VaadinFlowPluginExtension.kt:28]: Just commenting that there are a lot of things that are not configurable via maven here - are these configurations still valid for gradle ? Like deciding if bundle should be created or npm install run after updating the dependencies

**Review by pleku — CHANGES_REQUESTED** [review]
Couple of comments + questions

File: flow-plugins/flow-gradle-plugin/README.md (hunk 3 above)
Comment (caalador) [reply flow-plugins/flow-gradle-plugin/README.md:84]: I think it's from a time when npm was the default instead of pnpm so before v14.3?

File: flow-plugins/flow-gradle-plugin/build.gradle [hunk 10]
```diff
@@ -0,0 +1,221 @@
+/*
+ * Copyright 2018-2019 Devsoap Inc.
```
Comment (mstahv) [inline flow-plugins/flow-gradle-plugin/build.gradle:2]: I wonder what is left from the Devsoap version here in the build script that is worth copyright notice. Also, this is build script, maybe remove the license header altogether 🤔

File: flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/VaadinFlowPluginExtension.kt (hunk 9 above)
Comment (mstahv) [reply flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/VaadinFlowPluginExtension.kt:28]: @mvysny You might be the best source to comment.

File: flow-plugins/flow-gradle-plugin/README.md (hunk 3 above)
Comment (caalador) [reply flow-plugins/flow-gradle-plugin/README.md:84]: We should probably only have the development info in the README and the rest should live in the docs.

File: flow-plugins/flow-gradle-plugin/build.gradle (hunk 10 above)
Comment (pleku) [reply flow-plugins/flow-gradle-plugin/build.gradle:2]: I missed this, but nor do I understand how build scripts would even have license header (similar to pom.xml) and we'll just drop the header

File: flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/VaadinFlowPluginExtension.kt (hunk 9 above)
Comment (mvysny) [reply flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/VaadinFlowPluginExtension.kt:28]: Those settings were taken from Maven plugin 14.3.x's sources and they were most probably not cleaned up when adding support for Vaadin 20. We should find such settings and remove them, so that Gradle plugin has the same set of settings as Maven plugin does.

File: flow-plugins/flow-gradle-plugin/README.md (hunk 4 above)
Comment (mvysny) [reply flow-plugins/flow-gradle-plugin/README.md:89]: True, we should reword this to "...apart from the production mode, there's rarely any need to change..." or similar. The point was to stop users from tinkering with the settings, shooting themselves to the foot.

File: flow-plugins/flow-gradle-plugin/README.md (hunk 2 above)
Comment (mvysny) [reply flow-plugins/flow-gradle-plugin/README.md:64]: In such case I propose to add similar functionality to Maven plugin as requested here: https://github.com/vaadin/flow/issues/8885 . I have seen Vaadin Dance being helpful to users, and `vaadinClean` essentially performs the dance very quickly.

Comment (pleku) [reply flow-plugins/flow-gradle-plugin/README.md:64]: At least now the issue has a reference to here so it should be OK for now

File: flow-plugins/flow-gradle-plugin/README.md (hunk 5 above)
Comment (taefi) [reply flow-plugins/flow-gradle-plugin/README.md:218]: As the structure of gradle plugin was not changed during the process of adopting its source code under flow project, I believe this is still valid.

**Review by pleku — CHANGES_REQUESTED** [review]
Readme changes still missing, outdated content to be 🪓

File: flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/VaadinFlowPluginExtension.kt (hunk 9 above)
Comment (taefi) [reply flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/VaadinFlowPluginExtension.kt:28]: I scanned all the setting properties defined in `VaadinFlowPluginExtension.kt` and noticed all of them are present either in `FlowModeAbstractMojo.java` or in `BuildFrontendMojo.java`, especially `runNpmInstall` and `generateBundle`. Although there may be no way to configure them even in the maven and always their default values are in use, I propose to keep them for now, and clean them up in a separate task for both maven and gradle plugins at the same time.

File: flow-plugins/flow-gradle-plugin/build.gradle (hunk 10 above)
Comment (taefi) [reply flow-plugins/flow-gradle-plugin/build.gradle]: The license header is removed.

File: flow-plugins/flow-gradle-plugin/pom.xml (hunk 7 above)
Comment (taefi) [reply flow-plugins/flow-gradle-plugin/pom.xml:12]: Added.

File: flow-plugins/flow-gradle-plugin/README.md (hunk 6 above)
Comment (taefi) [reply flow-plugins/flow-gradle-plugin/README.md]: Done.

File: flow-plugins/flow-gradle-plugin/README.md (hunk 4 above)
Comment (taefi) [reply flow-plugins/flow-gradle-plugin/README.md]: `## Configuration` heading is removed from README and will be added to the docs.

File: flow-plugins/flow-gradle-plugin/README.md (hunk 3 above)
Comment (taefi) [reply flow-plugins/flow-gradle-plugin/README.md]: Removed from README.

File: flow-plugins/flow-gradle-plugin/README.md [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+This applies to Visual Studio Code (VSCode) as well since it also uses Eclipse bits and BuildShip
+underneath - see [https://github.com/mvysny/vaadin14-embedded-jetty-gradle/issues/4](https://github.com/mvysny/vaadin14-embedded-jetty-gradle/issues/4)
+for more details.
+
+In order to run your project in production mode from your IDE, simply compile the
+project in production mode, e.g. by using `-Pvaadin.productionMode`. The plugin will
+then produce the `META-INF/VAADIN/config/flow-build-info.json` file with the `"productionMode": true` setting,
+which will then cause Vaadin to start in production mode at runtime. To revert this setting
+back to false, simply recompile your project without the `-Pvaadin.productionMode` switch.
+
```
Comment (pleku) [inline flow-plugins/flow-gradle-plugin/README.md:147]: Several topics from above here are too not related to developing the plugin ... but you can revisit those after working on the docs

**Review by pleku — APPROVED** [review]
